cartographer/mapping/internal/2d/pose_graph_2d.cc

```diff
@@ -1237,6 +1237,26 @@ void PoseGraph2D::TrimmingHandle::TrimSubmap(const SubmapId& submap_id) {
     }
     parent_->data_.constraints = std::move(constraints);
   }
+
+  // A frozen trajectory needs special care because there we can't identify the
+  // nodes_to_remove based on optimization constraints (it doesn't have them).
+  if (parent_->IsTrajectoryFrozen(submap_id.trajectory_id)) {
+    // Remove node_ids that are associated with the frozen submap to be trimmed.
+    nodes_to_remove.insert(
+        parent_->data_.submap_data.at(submap_id).node_ids.begin(),
```

Comment:
Do these `node_ids` always exist? Would it then make sense to get rid of the old style of handling deletion of nodes and replace it by your improved strategy completely? Something along these lines:
1. Compute `nodes_to_retain` by going through all submaps and adding all `node_ids` for submaps not equal to the given `submap_id` to remove.
2. Compute `nodes_to_remove` as all `node_ids` of `submap_id` not to be retained.
3. Remove the constraints similar to before, but without the need to build up the `nodes_to_remove` as part of it. We could add a `CHECK` to validate that all retained constraints use nodes not in `nodes_to_remove`.

What do you think?

Reply:
Regarding your 3. point: There's already a loop that only keeps constraints for which `nodes_to_remove.count(constraint.node_id) == 0` holds. With this in place we don't need to check afterwards, no?